ports/geckolib/gecko_bindings/sugar/refptr.rs

```diff
@@ -0,0 +1,53 @@
+use std::marker::PhantomData;
+use std::mem::transmute;
+use std::ptr::null;
+use std::sync::Arc;
+
+pub unsafe trait RefCounted<'a> where Self: Sized {
+    type Strong;
+    type Borrowed;
+    fn into_strong(x: Arc<Self>) -> Self::Strong;
+    fn null_strong() -> Self::Strong;
+    fn from_borrowed(x: Self::Borrowed) -> &'a Arc<Self>;
+    fn into_borrowed(x: &'a Arc<Self>) -> Self::Borrowed;
+    fn from_borrowed_opt(x: Self::Borrowed) -> Option<&'a Arc<Self>> {
+        let arc = Self::from_borrowed(x);
+        let ptr: *const Self = &**arc as *const Self;
+        if ptr == null() {
+            None
+        } else {
+            Some(arc)
+        }
+    }
+    unsafe fn decref_borrowed(x: Self::Borrowed) where Self: 'a {
+        let _: Arc<Self> = transmute(Self::from_borrowed(x) as *const _);
```

Comment:
This looks wrong, as Self::from_borrowed(x) returns a &'a Arc<T>, and you're casting that reference into an Arc<T> - don't you want to dereference it first?